api/lql.go

```diff
@@ -18,24 +18,201 @@
+	// query
+	if err := q.TranslateQuery(); err != nil {
+		return err
+	}
+	// start
+	if start, err := q.TranslateTime(q.StartTimeRange); err != nil {
+		return err
+	} else {
+		q.StartTimeRange = start
+	}
+	// end
+	if end, err := q.TranslateTime(q.EndTimeRange); err != nil {
+		return nil
+	} else {
+		q.EndTimeRange = end
+	}
+	return nil
+}
+
+func (q *LQLQuery) TranslateQuery() error {
+	// empty
+	if q.QueryText != "" {
+		return nil
+	}
+	// json
+	var t LQLQuery
+
+	if err := json.Unmarshal([]byte(q.QueryBlob), &t); err == nil {
+		if q.StartTimeRange == "" {
+			q.StartTimeRange = t.StartTimeRange
+		}
+		if q.EndTimeRange == "" {
+			q.EndTimeRange = t.EndTimeRange
+		}
+		q.QueryText = t.QueryText
+		return err
+	}
+	// lql
+	if matched, _ := regexp.MatchString(reLQL, q.QueryBlob); matched {
+		q.QueryText = q.QueryBlob
+		return nil
+	}
+	return errors.New(LQLQueryTranslateError)
+}
+
+func (q LQLQuery) TranslateTime(inTime string) (outTime string, err error) {
+	// empty
+	if inTime == "" {
+		return
+	}
+	// parse time as RFC3339
+	if t, err := time.Parse(time.RFC3339, inTime); err == nil {
+		outTime = t.UTC().Format(time.RFC3339)
+		return outTime, err
+	}
+	// parse time as millis
+	if t, err := strconv.ParseInt(inTime, 10, 64); err == nil {
+		outTime = time.Unix(0, t*int64(time.Millisecond)).UTC().Format(time.RFC3339)
+		return outTime, err
+	}
+	err = errors.New("unable to parse time (" + inTime + ")")
+	return
+}
+
+func (q LQLQuery) ValidateRange(allowEmptyTimes bool) (err error) {
+	// validate start
+	var start time.Time
+	if q.StartTimeRange != "" {
+		if start, err = time.Parse(time.RFC3339, q.StartTimeRange); err != nil {
+			return err
+		}
+	} else if allowEmptyTimes {
+		start = time.Unix(0, 0)
+	} else {
+		return errors.New("start time must not be empty")
+	}
+	// validate end
+	var end time.Time
+	if q.EndTimeRange != "" {
+		if end, err = time.Parse(time.RFC3339, q.EndTimeRange); err != nil {
+			return err
+		}
+	} else if allowEmptyTimes {
+		end = time.Now()
+	} else {
+		return errors.New("end time must not be empty")
+	}
+	// validate range
+	if start.After(end) {
+		return errors.New("date range should have a start time before the end time")
+	}
+	return nil
+}
+
+type LQLQueryResponse struct {
+	Data    []LQLQuery `json:"data"`
+	Ok      bool       `json:"ok"`
+	Message string     `json:"message"`
+}
+
 // LQLService is a service that interacts with the LQL
 // endpoints from the Lacework Server
 type LQLService struct {
 	client *Client
 }
 
-type lqlQueryRequest struct {
-	QueryText string `json:"QUERY_TEXT"`
+func (svc *LQLService) CreateQuery(query string) (
+	response LQLQueryResponse,
+	err error,
+) {
+	lqlQuery := LQLQuery{QueryBlob: query}
+	if err = lqlQuery.Validate(true); err != nil {
+		return
+	}
+
+	err = svc.client.RequestEncoderDecoder("POST", ApiLQL, lqlQuery, &response)
+	return
+}
+
+func (svc *LQLService) GetQueries() (
+	response LQLQueryResponse,
+	err error,
+) {
```

Comment:
here you are not using "named returns", clean the function definition
```suggestion
func (svc *LQLService) GetQueries() (LQLQueryResponse, error) {
```
This is more readable 👆🏽 - right?

Reply:
Fixed.